### ISSUE TITLE
Cast Error when using SQL Server DistributedCache

### ISSUE THREAD

**Issue by monsieurberry** [PiranhaCMS/piranha.core#744]
Hi!

I'm just testing out the DistributedCache with SQL Server. I've created a Cache DB and table with ...

`dotnet sql-cache create "Data Source=(localdb)\MSSQLLocalDB;Initial Catalog=DistCache;Integrated Security=True;" dbo TestCache`

... and wired everything up in my Startup.cs by adding ...

`services.AddDistributedSqlServerCache(options =>`
`            {`
`                options.ConnectionString = Configuration["DistCache_ConnectionString"];`
`                options.SchemaName = "dbo";`
`                options.TableName = "TestCache";`
`            });`
`services.AddPiranhaDistributedCache();`

.. to ConfigureServices(), and ...

`App.CacheLevel = Piranha.Cache.CacheLevel.Full;`

to Configure().

When I first load up my site everything works perfectly. I check my cache table and everything is in there, so I know writing to the cache works and all is (hopefully) configured OK.

When I refresh my page (now getting from the cache) I get the following error...

> An unhandled exception occurred while processing the request.
InvalidCastException: Unable to cast object of type 'System.Collections.Generic.List`1[Piranha.Services.SiteService+SiteMapping]' to type 'System.String'.
Piranha.Cache.DistributedCache.Deserialize<T>(byte[] bytes) in DistributedCache.cs, line 128

I've tried debugging and a few different options, however I'm a bit lost - was hoping I could try and figure out what was going wrong but no luck unfortunately.

I also tried out the same on 6.1.0 but I got the same error.

If you need anymore info please let me know.

Thanks!

**Comment by monsieurberry** [comment]
Managed to fix the issue with the following changes to Deserialize<T> in DistributedCache.

A bit of a hack, but hopefully gives you guys an idea of where the problem is  :-)

> /// <summary>
        /// Deserializes the byte array to an object.
        /// </summary>
        /// <param name="bytes">The byte array</param>
        /// <typeparam name="T">The object type</typeparam>
        /// <returns>The deserialized object</returns>
        private T Deserialize<T>(byte[] bytes)
        {
            if (bytes == null)
            {
                return default(T);
            }

            var formatter = new BinaryFormatter();
            using (var stream = new MemoryStream(bytes))
            {
                var obj = formatter.Deserialize(stream);

                if (!(obj.GetType() == typeof(string)) && IsSerializable(obj.GetType()))
                {
                    return (T)obj;
                }
                else
                {
                    // First lets decode the byte array into a string
                    var json = (string)obj;

                    // Next deserialize the json into an object
                    var settings = new JsonSerializerSettings
                    {
                        TypeNameHandling = TypeNameHandling.All
                    };
                    return JsonConvert.DeserializeObject<T>(json, settings);
                }
            }
        }

**Comment by tidyui** [comment]
Thanks for investigating the issue further! We'll take a look at it.

Regards

**Comment by monsieurberry** [comment]
Thank you, Håkan!

I think I may have found a better solution for you.

It seems as if the SiteMapping is cached as a List object, but when it's retrieved from the cache in SiteService, it specifies IList...

I was able to solve the issue by changing IList<SiteMapping> to List<SiteMapping> in GetByHostnameAsync.

Hope this helps :-)

**Comment by nahojd** [comment]
I get (more or less) the same error when I try to use RedisCache (from Microsoft.Extensions.Caching.StackExchangeRedis) or even the default Distributed Memory Cache (services.AddDistributedMemoryCache), which is just an in-memory implementation. 

Specifically, I get

```text
An exception of type 'System.InvalidCastException' occurred in System.Private.CoreLib.dll but was not handled in user code: 'Unable to cast object of type 'System.Collections.Generic.List`1[Piranha.Models.PostType]' to type 'System.String'.'
   at Piranha.Cache.DistributedCache.Deserialize[T](Byte[] bytes)
   at Piranha.Cache.DistributedCache.Get[T](String key)
   at Piranha.Services.PostTypeService.<GetTypes>d__8.MoveNext()
```

or sometimes in `PageTypeService.GetTypes` instead.

When i use RedisCache I can see that the data has been written to Redis, but it can't retrieve it. 

Has anyone got DistributedCache working?

**Comment by tidyui** [comment]
I will see if I can replicate the error on the `version/v8.1` branch.

**Comment by tidyui** [comment]
So I have looked into the problem and apparently how models are loaded have changed a bit since the distributed cache was written. Also, unfortunately we don't run any integration tests on distributed cache, only on memory cache which I will look into, otherwise we would've caught the problem earlier.

For those of you that need distributed caching while we try to resolve the issue, I've created a gist that contains an alternate distributed cache that doesn't use binary formatting, instead it serializes objects using `Newtonsoft.Json`. This will not be as fast as using a binary formatter, but it's a good intermediate solution while waiting for the official fix.

https://gist.github.com/tidyui/95c441293af55f9eb906ba756ae398e5

Best regards
